README.md
# geostools
An R package of tools to assist with geoscientific problems.


NEWS.md
# geostools 0.2.0

Changes to narrow the dependencies of **geotools**:

* Function `MinimizeSpherical()` is copied from package **pfields** in order to
  be able to use the full functionality of the function `GetDistance()` without
  the need to have a suggest dependence on **pfields** (in turn, the function
  can be removed from **pfields** to have **pfields** depend on
  **geotools**). This also offers a clearer thematic distinction between the two
  packages.
* There is no need to import base package **Matrix** into the NAMESPACE, a
  simple package dependence is enough.
  
In addition, some improvements on function documentation are added.

# geostools 0.1.0

* Initial package version with number and scope of the functions incl. their
  documentation identical to their original versions in deprecated package
  'ecustools' which they were a part of ('ecustools'
  [`main`](https://github.com/EarthSystemDiagnostics/ecustools/tree/master)
  branch as of 2020-11-20).


R/MinimizeSpherical.R
#' Minimize difference between coordinates
#'
#' This function finds the position in a latitude/longitude field that is
#' closest to a given target position by minimizing the root mean square
#' deviation of their position vectors in spherical coordinates.
#'
#' @param lat0 a length-1 vector with the latitude [degree] of the target
#'   position.
#' @param lon0 a length-1 vector with the longitude [degree] of the target
#'   position.
#' @param lat numeric vector of latitudes [degree] from a field of coordinates
#'   for which the point closest to the target is to be determined.
#' @param lon numeric vector of longitudes [degree] from a field of coordinates
#'   for which the point closest to the target is to be determined. Must have
#'   the same length as \code{lat}.
#' @param return.coordinates logical; shall the actual coordinates of the
#'   position closest to the target be returned? Defaults to \code{FALSE}, which
#'   returns just the vector index of the closest position.
#' @return per default, the index position in \code{lat} (or \code{lon}) of the
#'   position closest to the target; or the actual coordinates of this position
#'   as a length-2 numeric vector (for \code{return.coordinates = TRUE}).
#' @author Thomas Münch
#' @examples
#' # some coordinates
#' lat0 <- -75
#' lon0 <- 0
#' lat <- seq(-74, -80, -2)
#' lon <- seq(-2, 7, 3)
#'
#' # get closest point to lat0/lon0 in degree
#' MinimizeSpherical(lat0, lon0, lat, lon,
#'                   return.coordinates = TRUE)
#' @export
MinimizeSpherical <- function(lat0, lon0, lat, lon,
                              return.coordinates = FALSE) {

  rmsd <- function(v1, v2) {
    sqrt(mean((v1 - v2)^2, na.rm = TRUE))
  }

  # Error checking

  if (length(lat0) != length(lon0))
    stop("'lat0' and 'lon0' must both have length 1.")

  if (length(lat) != length(lon))
    stop("'lat' and 'lon' must have equal length.")

  n <- length(lat)

  # Convert angles from degree to radian

  x <- deg2rad(lat0)
  y <- deg2rad(lon0)

  xx <- deg2rad(lat)
  yy <- deg2rad(lon)

  # Convert input to vectors in spherical coordinates

  target <- c(cos(x) * cos(y),
              cos(x) * sin(y),
              sin(x))

  field.vectors <- rbind(cos(xx) * cos(yy),
                         cos(xx) * sin(yy),
                         sin(xx))

  # Root mean square deviation between target and data position vectors

  dev <- rep(NA, n)
  for (i in 1 : n) {

    dev[i] <- rmsd(field.vectors[, i], target)
  }

  # Return index of the position vector with minimum deviation to target

  i <- which.min(dev)

  if (return.coordinates) {
    res <- c(lat = lat[i], lon = lon[i])
  } else {
    res <- i
  }

  return(res)
}


R/interp.barycentric.R
#' ND barycentric interpolation at points Xi for a function with values f
#' measured at locations X
#'
#' @param X An n-by-d matrix. The rows of x represent n points in d-dimensional
#'   space.
#' @param f A vector of values nrow(X) long to be interpolated.
#' @param Xi An ni-by-d matrix. The rows of xi represent n points in
#'   d-dimensional space whose positions in the mesh are being sought.
#' @references Adapted from code here:
#'   https://dahtah.wordpress.com/2013/03/06/barycentric-interpolation-fast-interpolation-on-arbitrary-grids/
#'
#' @author Andrew Dolman <andrew.dolman@awi.de>
#' @return matrix
#' @export
#'
#' @examples
#' n <- 100
#' df <- data.frame(x = runif(n, -1, 1), y = runif(n, -1, 1), z = runif(n, -1, 1))
#' df$v <- with(df, 2*x + 1*y^2 + 3*z)
#' df.reg <- with(df, expand.grid(x = seq(min(x), max(x), length.out = 5),
#'                                y = seq(min(y), max(y), length.out = 5),
#'                                z = seq(min(z), max(z), length.out = 5)
#' ))
#'
#' out <- InterpBarycentric(X = df[,c(1,2,3)],
#'                           f = df$v,
#'                           Xi = df.reg)
#'
#' out <- as.data.frame(out)
#'
#' out$v_true <-  with(out, 2*x + 1*y^2 + 3*z)
#'
#' plot(v~v_true, data = out)
#' abline(0, 1)
#'
#' hist(out$v - out$v_true)
InterpBarycentric <- function(X, f, Xi) {
  X <- as.matrix(X)
  Xi <- as.matrix(Xi)
  dn <- geometry::delaunayn(X)
  tri <- geometry::tsearchn(X, dn, Xi, fast = TRUE)

  # points in Xi that fall inside polygons
  good.ind <- which(is.na(tri$idx) == FALSE)
  Xi.2 <- Xi[good.ind, ]
  tri <- geometry::tsearchn(X, dn, Xi.2, fast = TRUE)

  # For each line in Xi.2, defines which points in X contribute to the
  # interpolation
  active <- dn[tri$idx, ]
  # Define the interpolation as a sparse matrix operation. Faster than using apply,
  # probably slower than a C implementation
  M <- Matrix::sparseMatrix(i = rep(1:nrow(Xi.2), each = ncol(Xi.2) + 1), j = as.numeric(t(active)),
                            x = as.numeric(t(tri$p)), dims = c(nrow(Xi.2), length(f)))
  v <- as.numeric(M %*% f)

  out <- cbind(Xi, v = NA)
  out[good.ind, "v"] <- v
  return(out)
}


R/GetDistance.R
#' Find distance between points on Earth
#'
#' This is a general function to calculate the distance on Earth between a
#' target location and all points in a field of coordinates. It provides a
#' straighforward interface to the function \code{\link[geosphere]{distGeo}}
#' from package \code{geosphere}.
#'
#' This is a wrapper function whose purpose it is to provide a straighforward
#' interface to the function \code{\link[geosphere]{distGeo}}; this includes:
#' \itemize{
#' \item direct input of latitude/longitude values which are then shaped into
#'       the appropriate structure for \code{\link[geosphere]{distGeo}};
#' \item longitude values are automatically confined to [-180, 180] as needed by
#'       \code{\link[geosphere]{distGeo}};
#' \item if required, one can select the point closest to the target location as
#'       the reference for the distance calculations.}
#' @param lat0 numeric vector of length 1 with the latitude (in degree) of the
#' target location relative to which distances are to be calculated.
#' @param lon0 numeric vector of length 1 with the longitude (in degree) of the
#' target location.
#' @param lat numeric vector of latitudes (in degree) of the positions for
#' which distances from the target position are to be calculated.
#' @param lon numeric vector of longitudes (in degree) of the positions for
#' which distances from the target position are to be calculated. Must have the
#' same length as \code{lat}.
#' @param get.nearest logical; specify whether the given 'lat0' - 'lon0' pair
#' shall be used _as is_ as the target location (\code{FALSE}), or instead the
#' nearest point of the positions specified by 'lat' and 'lon' (\code{TRUE}).
#' Selection of the nearest point is currently implemented with
#' \code{\link{MinimizeSpherical}}.
#' @param verbose logical; if \code{TRUE} an informative message regarding the
#' selected target location is printed.
#' @return Numeric vector of the same length as \code{lat} with the distances
#' in km from the target location.
#' @author Thomas Münch
#' @seealso \code{\link{MinimizeSpherical}}
#' @examples
#' # some coordinates
#' lat0 <- -75
#' lon0 <- 0
#' lat <- seq(-74, -80, -2)
#' lon <- seq(-2, 7, 3)
#'
#' d <- GetDistance(lat0, lon0, lat, lon, verbose = TRUE)
#' range(d)
#'
#' d <- GetDistance(lat0, lon0, lat, lon, get.nearest = TRUE, verbose = TRUE)
#' range(d)
#' @export
GetDistance <- function(lat0, lon0, lat, lon,
                        get.nearest = FALSE, verbose = FALSE) {

  # Error checking

  if (length(lat0) != length(lon0))
    stop("'lat0' and 'lon0' must both have length 1.")

  if (length(lat) != length(lon))
    stop("'lat' and 'lon' must have equal length.")

  # Select target point based on input

  if (get.nearest) {

    # get lat, lon of nearest point relative to target
    point <- rev(MinimizeSpherical(lat0, lon0, lat, lon,
                                   return.coordinates = TRUE))
    
    if (verbose) {
      message(sprintf(
        "Nearest grid point used as target: lat0 = %f, lon0 = %f.",
        point[2], point[1]))
    }

  } else {

    if (verbose) message("Using supplied 'lat0' and 'lon0' value as target.")
    point <- c(lon0, lat0)
  }

  # Shape coordinates of field in appropriate structure for 'distGeo' function

  field.coord <- rbind(lon, lat)

  # Need to put longitudes in [-180, 180] for 'distGeo' function

  point[point >= 180] <- point[point >= 180] - 360
  field.coord[field.coord >= 180] <- field.coord[field.coord >= 180] - 360

  # Get distance field
  distance <- apply(field.coord, 2, geosphere::distGeo, p2 = point)

  # Convert to km
  distance <- distance / 10^3

  return(distance)
}
